### PR TITLE
[Merged by Bors] - feat(LinearAlgebra, Cardinal): new cardinal lemmas to generalize some results about `Module.rank` in #9151

### PR DESCRIPTION
**Pull request by alreadydone**
+ Proves that `Sup` (ciSup) commutes with cardinal addition (`ciSup_add_ciSup`) and multiplication. Generalize results in Cardinal/Basic introduced in #8842 to achieve this.

+ Use `ciSup_add_ciSup` to prove that the rank of a module is always at least the rank of a submodule plus the rank of the quotient by the submodule. Deduce that the rank of a product module is at least the sum of the ranks of the two factors.

+ Show that quotienting by a torsion submodule preserves the rank.

+ Golf `rank_zero_iff_forall_zero` using a recently added lemma.

Co-authored-by: Andrew Yang <the.erd.one@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 1]
```diff
@@ -108,6 +108,9 @@ protected irreducible_def Module.rank : Cardinal :=
   ⨆ ι : { s : Set V // LinearIndependent K ((↑) : s → V) }, (#ι.1)
 #align module.rank Module.rank
 
+instance : Nonempty { s : Set V // LinearIndependent K ((↑) : s → V) } :=
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Dimension.lean:111]: Can you add a comment explaining why this instance is useful? I don't think it is particularly reasonably to add such an instance for every predicate that could ever exist!

**Comment by alreadydone** [comment]
!bench

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 2]
```diff
@@ -108,6 +108,9 @@ protected irreducible_def Module.rank : Cardinal :=
   ⨆ ι : { s : Set V // LinearIndependent K ((↑) : s → V) }, (#ι.1)
 #align module.rank Module.rank
 
+lemma nonempty_linearIndependent_set : Nonempty {s : Set V // LinearIndependent K ((↑) : s → V)} :=
```
Comment (erdOne) [inline Mathlib/LinearAlgebra/Dimension.lean:111]: Should this not be an instance?

Comment (erdOne) [reply Mathlib/LinearAlgebra/Dimension.lean:111]: Oh. For performance reasons?

**Comment by erdOne** [comment]
!bench

File: Mathlib/SetTheory/Cardinal/Ordinal.lean [hunk 3]
```diff
@@ -849,6 +849,54 @@ protected theorem eq_of_add_eq_add_right {a b c : Cardinal} (h : a + b = c + b)
...
+
+protected theorem add_ciSup : c + (⨆ i, f i) = ⨆ i, c + f i := by
+  rw [add_comm, Cardinal.ciSup_add f hf]; simp_rw [add_comm]
+
+protected theorem ciSup_add_ciSup (g : ι' → Cardinal.{v}) (hg : BddAbove (range g)) :
+    (⨆ i, f i) + (⨆ j, g j) = ⨆ (i) (j), f i + g j := by
+  simp_rw [Cardinal.ciSup_add f hf, Cardinal.add_ciSup g hg]
+
+protected theorem ciSup_mul : (⨆ i, f i) * c = ⨆ i, f i * c := by
+  have : ∀ i, f i * c ≤ (⨆ i, f i) * c := fun i ↦ mul_le_mul_right' (le_ciSup hf i) c
```
Comment (erdOne) [inline Mathlib/SetTheory/Cardinal/Ordinal.lean:877]:
```suggestion
protected theorem ciSup_mul : (⨆ i, f i) * c = ⨆ i, f i * c := by
  by_cases hf : BddAbove (range f); swap
  · by_cases hc : c = 0; · simp [hc]
    show (dite _ _ _) * c = dite _ _ _
    rw [dif_neg, dif_neg, bot_eq_zero', zero_mul]
    · exact hf
    · exact fun hx ↦ hf ⟨⨆ i, f i * c, fun _ ⟨i, hi⟩ ↦ (le_mul_right hc).trans (hi ▸ le_ciSup hx i)⟩
  have : ∀ i, f i * c ≤ (⨆ i, f i) * c := fun i ↦ mul_le_mul_right' (le_ciSup hf i) c
```
This removes the `BddAbove` hypothesis.

Comment (alreadydone) [reply Mathlib/SetTheory/Cardinal/Ordinal.lean:877]: Thanks! I mostly copied the proof from `ciSup_add` but didn't check whether the hypotheses remain necessary. I think `Nonempty` can also be removed.

File: Mathlib/LinearAlgebra/Dimension.lean [hunk 4]
```diff
@@ -241,6 +263,25 @@ theorem rank_quotient_le (p : Submodule R M) : Module.rank R (M ⧸ p) ≤ Modul
 
 variable [Nontrivial R]
 
+variable (R M M')
+
+open LinearMap in
+theorem lift_rank_add_lift_rank_le_rank_prod [Nontrivial R] :
```
Comment (erdOne) [inline Mathlib/LinearAlgebra/Dimension.lean:269]:
```suggestion
theorem lift_rank_add_lift_rank_le_rank_prod :
```
This is already in variables

**Comment by erdOne** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**